### PR TITLE
add reducer to app

### PR DESCRIPTION
**Pull request by KaterynaMadzhara**

### CODE REVIEW

File: todo_react/todo-react/src/models.ts [hunk 1]
```diff
@@ -1,5 +1,12 @@
...
+    completed?: boolean
     title: string
+}
+
+export interface IAction {
+    type?: string
+    id?: number
+    completed?: boolean
+    title: string|""
 }
```
Comment (Romms) [inline todo_react/todo-react/src/models.ts]: i would recommend to use type unions to have different types depend on action type

File: todo_react/todo-react/src/components/AddToDo.tsx [hunk 2]
```diff
@@ -1,21 +1,36 @@
...
+    const onChange = (event: ChangeEvent<HTMLInputElement>) => {
+        setCurrentTitle(event.target.value)
+    }
+    const onClick = (event: any) => {
+        event.preventDefault()
+        if (!currentTitle || !currentTitle.trim()) {
+            setError("Please, enter valid value")
+            return
+        }
+        dispatch({type:"added", id: Math.floor(Math.random() * 100), completed: false, title: currentTitle})
```
Comment (gerasimvol) [inline todo_react/todo-react/src/components/AddToDo.tsx]: you can move `id` creation/calculation into reducer, to make `dispatch` calls cleaner (pass only todo related data. `id` is more like metadata)

File: todo_react/todo-react/src/context/contextProvider.tsx [hunk 3]
```diff
@@ -1,84 +1,68 @@
...
+            }
+            default: {
+                throw Error('Unknown action: ' + action.type);
+            }
+        }
     }
+    const initialTodo = JSON.parse(localStorage.getItem('TodoList') || '[]')
+    const [todos, dispatch] = useReducer(
+        tasksReducer,
+        initialTodo);
```
Comment (Romms) [inline todo_react/todo-react/src/context/contextProvider.tsx]: Just an idea - add todos to deps to a useEffect to update local store
```
    useEffect(() => {
      localStorage.setItem('TodoList', JSON.stringify(todos))
    }, [todos])
```

File: todo_react/todo-react/src/models.ts [hunk 4]
```diff
@@ -1,5 +1,12 @@
 export interface ITodo {
-    id: number
-    completed: boolean
+    id?: number
+    completed?: boolean
     title: string
+}
+
+export interface IAction {
+    type?: string
```
Comment (gerasimvol) [inline todo_react/todo-react/src/models.ts]: `type` is required

File: todo_react/todo-react/src/models.ts [hunk 5]
```diff
@@ -1,5 +1,12 @@
...
+    id?: number
+    completed?: boolean
+    title: string|""
```
Comment (gerasimvol) [inline todo_react/todo-react/src/models.ts]: empty string `""` is string anyway

File: todo_react/todo-react/src/models.ts [hunk 6]
```diff
@@ -1,5 +1,12 @@
 export interface ITodo {
-    id: number
-    completed: boolean
+    id?: number
```
Comment (gerasimvol) [inline todo_react/todo-react/src/models.ts]: how Todo can have no `id`?

File: todo_react/todo-react/src/context/contextProvider.tsx [hunk 7]
```diff
@@ -1,84 +1,68 @@
...
-        const newTodos = todos.map(todoEl => {
-            let newEl = {...todoEl}
-            if (todoId === newEl.id) {
-                newEl.completed = !newEl.completed
+
+    const tasksReducer = (todos: ITodo[], action: IAction) => {
+        switch (action.type) {
+            case 'added': {
+                let newTodos = [...todos, {
+                    id: action.id,
```
Comment (gerasimvol) [inline todo_react/todo-react/src/context/contextProvider.tsx]: you can create `id` here. not pass as param

File: todo_react/todo-react/src/context/contextProvider.tsx [hunk 8]
```diff
@@ -1,84 +1,68 @@
...
-        const newTodos = todos.filter(todoEl => todoId !== todoEl.id)
-        saveToDos(newTodos)
-    }
-    const doneTodo = (todoId: number) => {
-        const newTodos = todos.map(todoEl => {
-            let newEl = {...todoEl}
-            if (todoId === newEl.id) {
-                newEl.completed = !newEl.completed
+
+    const tasksReducer = (todos: ITodo[], action: IAction) => {
```
Comment (gerasimvol) [inline todo_react/todo-react/src/context/contextProvider.tsx]: I saw `todo`, `task`, `ToDo` etc.
Make give this entity common name

File: todo_react/todo-react/src/context/contextProvider.tsx [hunk 9]
```diff
@@ -1,84 +1,68 @@
...
-            setError("Please, enter valid value")
-            return
-        }
-        addTodo({id: Math.floor(Math.random() * 100), completed: false, title: currentTitle})
-        setCurrentTitle("")
-        setError("")
-    }
+    useEffect(() => {
+        window.addEventListener('storage', () => {
+            console.log("change to local storage!");
```
Comment (gerasimvol) [inline todo_react/todo-react/src/context/contextProvider.tsx]: avoid console.logs (debug stuff) in PRs

**Review by Romms — APPROVED** [review]

File: todo_react/todo-react/src/App.css [hunk 10]
```diff
@@ -73,12 +73,12 @@ h2 {
     display: flex;
 }
 
-#new-task-form {
+.new-task-form {
```
Comment (gerasimvol) [inline todo_react/todo-react/src/App.css]: task

File: todo_react/todo-react/src/models.ts [hunk 11]
```diff
@@ -2,4 +2,11 @@ export interface ITodo {
     id: number
     completed: boolean
     title: string
+}
+
+export interface IAction {
+    type: string
+    id?: number
```
Comment (gerasimvol) [inline todo_react/todo-react/src/models.ts:9]: how `id` can be optional? every todo item required to have an `id`

File: todo_react/todo-react/src/models.ts [hunk 12]
```diff
@@ -2,4 +2,11 @@ export interface ITodo {
...
+    completed?: boolean
```
Comment (gerasimvol) [inline todo_react/todo-react/src/models.ts:10]: Actually all fields are required

`completed` by default `false`
`title` by default empty `string`

why optional?

File: todo_react/todo-react/src/context/contextProvider.tsx [hunk 13]
```diff
@@ -1,81 +1,63 @@
...
+            case 'localstorage_changed': {
+                return JSON.parse(localStorage.getItem('TodoList') || '[]');
+            }
+            default: {
+                throw Error('Unknown action: ' + action.type);
+            }
+        }
     }
-    const contextValues: TodosContextModel = {
+    const initialTodo = JSON.parse(localStorage.getItem('TodoList') || '[]');
```
Comment (gerasimvol) [inline todo_react/todo-react/src/context/contextProvider.tsx]: `initialTodos` 
its plural (multiple items)

**Pull request merged**